### PR TITLE
Add kernel regression tests

### LINKED ISSUE #2747: Enable backward compat tests
### Describe the feature you'd like supported

To ensure eBPF platform is backward compatible, we need to enable tests for backward compatibility. Currently xdp-for-windows uses the test binaries from a previous release to execute on the latest bits, to ensure there is no regression.

We cannot do the same as most of our tests are unit_tests, which means they statically compile with the ebpfcore, api, etc. So there is no way to run the older tests against new binaries.

Currently there is a minimal test which loads a 0.9 native driver using latest drivers.

### Proposed solution

The solution has basically 2 parts:

**Part 1**
We currently have some kernel tests. Automate the existing kernel tests for regression tests -- test binaries from previous releases should pass using the latest eBPF binaries. This can possibly be run as scheduled task.

**Part 2**
Refactor all the unit tests which use public eBPF APIs so that they can be compiled both for unit tests and for kernel tests.
Once this is done, automate these new kernel tests also for regression tests.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by saxena-anurag**
Fixes #2747
## Description

This PR adds new kernel regression tests. The regression tests run currently on WS2022 for Debug and Release flavors.

Regression tests download artifacts from a previous release and use the following from the previous release:
1. Native modules for all samples -- this ensures NMR interface for native modules is backward compatible.
2. Extensions (netebpfext, samplext) -- this ensures NMR interface for extensions is backward compatible.
3. All the test exe -- this ensures all the eBPF APIs are backward compatible.

## Testing

This PR adds tests.

## Documentation

No.

## Installation

No.

### CODE REVIEW

File: scripts/config_test_vm.psm1 [hunk 1]
```diff
@@ -576,6 +579,56 @@ function Get-RegressionTestArtifacts
     }
 }
 
+function Get-RegressionTestArtifacts
+{
+    param([Parameter(Mandatory=$True)] $Configuration)
+
+    $ArtifactVersion = "0.15.0"
```
Comment (shankarseal) [inline scripts/config_test_vm.psm1:586]: If you parameterize this, then in future we can control this from the yml.

File: scripts/config_test_vm.psm1 [hunk 2]
```diff
@@ -576,6 +579,56 @@ function Get-RegressionTestArtifacts
...
+    # Verify artifacts' folder presence
+    if (-not (Test-Path -Path $RegressionTestArtifactsPath)) {
+        $ErrorMessage = "*** ERROR *** Regression test artifacts folder not found: $RegressionTestArtifactsPath)"
+        Write-Log $ErrorMessage
+        throw $ErrorMessage
+    }
+
+    # Download regression test artifacts for each version.
+    $DownloadPath = "$RegressionTestArtifactsPath"
+    # $ArtifactName = "v$ArtifactVersion/Build-x64-native-only-Release.$ArtifactVersion.zip"
```
Comment (shankarseal) [inline scripts/config_test_vm.psm1:603]: delete.

File: scripts/config_test_vm.psm1 [hunk 3]
```diff
@@ -576,6 +579,56 @@ function Get-RegressionTestArtifacts
...
+    $ArtifactName = "Release-v$ArtifactVersion/Build-x64.$Configuration.zip"
+    $ArtifactUrl = "https://github.com/microsoft/ebpf-for-windows/releases/download/" + $ArtifactName
+
+    Write-Log "Downloading regression test artifacts for version $ArtifactVersion" -ForegroundColor Green
+    $ProgressPreference = 'SilentlyContinue'
+    Invoke-WebRequest -Uri $ArtifactUrl -OutFile "$DownloadPath\artifact.zip"
+
+    Write-Log "Extracting $ArtifactName"
+    Expand-Archive -Path "$DownloadPath\artifact.zip" -DestinationPath $DownloadPath -Force
+    # Expand-Archive -Path "$DownloadPath\build-NativeOnlyRelease.zip" -DestinationPath $DownloadPath -Force
```
Comment (shankarseal) [inline scripts/config_test_vm.psm1:613]: delete all commented out code.

File: tests/api_test/api_test.cpp [hunk 4]
```diff
@@ -1311,6 +1311,84 @@ TEST_CASE("test_ringbuffer_wraparound", "[stress]")
     // Unsubscribe from the ring buffer.
     ring_buffer__free(ring);
 
+    // Clean up.
+    bpf_object__close(object);
+}
+
+TEST_CASE("anusa_new_test", "[anusa_new_test]")
```
Comment (shankarseal) [inline tests/api_test/api_test.cpp]: please rename test case.

Comment (shankarseal) [inline tests/api_test/api_test.cpp]: what is the utility of this new test?

File: scripts/install_ebpf.psm1 [hunk 5]
```diff
@@ -239,6 +244,64 @@ function Install-eBPFComponents
...
+                            Write-Log("$($_.Key) service started.") -ForegroundColor Green
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+
+    # Copy ebpfapi.dll from installation directory to current directory.
```
Comment (shankarseal) [inline scripts/install_ebpf.psm1:299]: why is this needed? Why can there be more than one versions of ebpfapi on the system?

File: tests/api_test/api_test.cpp (hunk 4 above)
Comment (saxena-anurag) [reply tests/api_test/api_test.cpp]: This has been removed in the latest commit. I had added this to validate the regression tests.

File: scripts/install_ebpf.psm1 (hunk 5 above)
Comment (saxena-anurag) [reply scripts/install_ebpf.psm1:299]: There will be 2 copies of ebpfapi.dll (and this behavior exists prior to this PR too) -- one in the artifacts folder, and another is added in `programfiles\ebpf-for-windows` by MSI installation.

File: tests/api_test/api_test.cpp (hunk 4 above)
Comment (saxena-anurag) [reply tests/api_test/api_test.cpp]: This has been removed.

File: scripts/install_ebpf.psm1 (hunk 5 above)
Comment (saxena-anurag) [reply scripts/install_ebpf.psm1:299]: TODO: File a bug that ebpfapi.dll is not in path

File: .github/workflows/cicd.yml [hunk 6]
```diff
@@ -251,6 +251,25 @@ jobs:
...
 
+  # Run the regression driver tests on self-hosted runners (only for 2022).
+  regression_driver_ws2022:
+    # Always run this job.
+    # Only run this on repos that have self-host runners.
+    needs: regular
+    if: github.repository == 'microsoft/ebpf-for-windows' && (github.event_name == 'schedule' || github.event_name == 'pull_request' || github.event_name == 'push' || github.event_name == 'merge_group' || github.event_name == 'workflow_dispatch')
+    uses: ./.github/workflows/reusable-test.yml
+    with:
+      pre_test: .\setup_ebpf_cicd_tests.ps1 -KmTracing $true -KmTraceType "file" -TestMode "Regression" -RegressionArtifactsVersion "0.15.0"
```
Comment (matthewige) [inline .github/workflows/cicd.yml:262]: When would we update this version? Does this get updated on every release? Or would we only update this when we have new breaking changes?

**Review by matthewige — APPROVED** [review]

File: .github/workflows/cicd.yml (hunk 6 above)
Comment (mtfriesen) [reply .github/workflows/cicd.yml:262]: Ideally the downlevel tests would be a matrix of supported versions.

**Review by mtfriesen — APPROVED** [review]

File: scripts/install_ebpf.psm1 (hunk 5 above)
Comment (gtrevi) [reply scripts/install_ebpf.psm1:299]: If I interpreted the intention correctly from the comment, we want to use the latest API version. If so, then instead of copying it and adding it to the path, why not just delete it from the regression artifacts, and have the DLL automatically resolved to the installation path?

File: .github/workflows/cicd.yml (hunk 6 above)
Comment (shankarseal) [reply .github/workflows/cicd.yml:262]: This is now https://github.com/microsoft/ebpf-for-windows/issues/3424.

File: scripts/install_ebpf.psm1 (hunk 5 above)
Comment (shankarseal) [reply scripts/install_ebpf.psm1:299]: The intention is not to have two copies of ebpfapi.dll. There shall be only one copy and that belongs to the build under test. But apparently, per Anurag, the tests are failing unless ebpfapi.dll is not copied in the local directory as the legacy test binaries. This seems like a bug which I am tracking using https://github.com/microsoft/ebpf-for-windows/issues/3425.

File: .github/workflows/cicd.yml (hunk 6 above)
Comment (shankarseal) [reply .github/workflows/cicd.yml:262]: Tracked using https://github.com/microsoft/ebpf-for-windows/issues/3424
